timApp/modules/cs/js/editor/math-editor/formula-editor.component.ts

```diff
@@ -0,0 +1,280 @@
+/**
+ * Formula Editor for inputting LaTeX math
+ * @author Juha Reinikainen
+ * @licence MIT
+ * @date 28.2.2023
+ */
+
+import {
+    Component,
+    ElementRef,
+    EventEmitter,
+    Input,
+    Output,
+    ViewChild,
+} from "@angular/core";
+import {FormControl} from "@angular/forms";
+import type {
+    IMathQuill,
+    MathFieldMethods,
+    MathQuillConfig,
+} from "vendor/mathquill/mathquill";
+import {IEditor} from "../editor";
+
+/**
+ * Field which has the focus
+ */
+enum ActiveEditorType {
+    Visual = "visual",
+    Latex = "latex",
+}
+
+/**
+ * editor.content split at current cursor location in editor
+ */
+type BeforeAndAfter = {
+    before: string;
+    after: string;
+};
+
+/**
+ * OldContent is split into two at cursor location if insert operation is supported
+ * or else just the content: string
+ */
+type OldContent = BeforeAndAfter | string;
+
+@Component({
+    selector: "cs-formula-editor",
+    template: `
+        <div [hidden]="!visible" class="formula-editor">
+            <div class="formula-editor-dialog">
+                <div class="formula-container">
+                    <span class="visual-input" #visualInput></span>
+        
+                    <textarea name="math-editor-output" #latexInput cols="30" rows="5"
+                              (click)="handleLatexFocus()"
+                              (keyup)="handleLatexInput()"
+                              [formControl]="latexInputControl"
+                              placeholder="Write LaTeX" i18n-placeholder>
+                    </textarea>                    
+                </div>
+        
+                <div class="formula-button-container">
+                    <div class="formula-buttons">
+                        <button class="timButton" (click)="handleFormulaOk()">Ok</button>
+                        <button class="timButton" (click)="handleFormulaCancel()" i18n>Cancel</button>                                        
+                    </div>
+        
+                    <label class="font-weight-normal">
+                        <input type="checkbox" [formControl]="isMultilineFormulaControl"><ng-container i18n>
+                        Multiline</ng-container>
+                    </label>
+                </div>
+            </div>
+        </div>
+            
+    `,
+    styleUrls: ["./formula-editor.component.scss"],
+})
+export class FormulaEditorComponent {
+    latexInputControl = new FormControl("");
+    @ViewChild("latexInput") latexInput!: ElementRef<HTMLTextAreaElement>;
+
+    @ViewChild("visualInput") visualInput!: ElementRef<HTMLElement>;
+    MQ!: IMathQuill;
+
+    mathField!: MathFieldMethods;
+
+    activeEditor: ActiveEditorType = ActiveEditorType.Visual;
+
+    oldContent!: OldContent;
+
+    @Output() okEvent = new EventEmitter<void>();
+    @Output() cancelEvent = new EventEmitter<void>();
+
+    isMultilineFormulaControl = new FormControl(true);
```

Comment:
Samoin kuin `latexInputControl`, eli voi käyttää `ngModel`ia sekä yksinkertaista booleania. Lisätietoja ja esimerkki: <https://angular.io/api/forms/NgModel#using-ngmodel-on-a-standalone-control>

Reply:
Miten tämän saa sitten toimimaan tavallisella booleanilla?

this.isMultilineFormulaControl.valueChanges.subscribe((value) => {
            this.updateFormulaToEditor();
        });

Reply:
Varmaan sillee set get tyylillä